### ISSUE TITLE
inefficient _validate_input and mistake

### ISSUE THREAD

**Issue by Trotew** [patrick-kidger/torchcubicspline#12]
The function `_validate_input` seems to be incorrect. The code tries to block non-monotonic data, but since `prev_t_i` is not updated, it doesn't seem to work. And in a Cuda environment, this part of the code is very inefficient.

**Comment by patrick-kidger** [comment]
Good catch. I'd be happy to accept a PR fixing this.

Yep, none of this code is very efficient on GPU. My own use case for this was interpolating data as a preprocessing step prior to training, so it didn't matter.

If you're agnostic to the choice of autodiff framework, then you may like to have a look at [Diffrax](https://github.com/patrick-kidger/diffrax), which includes code for [backward Hermite cubic splines](https://docs.kidger.site/diffrax/api/interpolation/#diffrax.backward_hermite_coefficients). (Rather than the natural cubic splines in this package.) Unlike the code here that should be relatively GPU-efficient.

**Comment by Trotew** [comment]
Thanks for the recommendation. I actually designed an interpolation module in a PyTorch model that needs to run online, so I am looking for a more efficient open-source PyTorch-based implementation of cubic spline interpolation.

I am also currently considering whether to switch to a more efficient interpolation method or to rewrite a more efficient implementation.

**Comment by patrick-kidger** [comment]
Right. If you need it to run online then the implementation in this repository won't be suitable, I'm afraid, as natural cubic splines don't satisfy that property: the "future" affects the "past".

The backward Hermite cubic splines I mentioned above are probably the appropriate algorithmic tool here, but of course if you're constrained to PyTorch then you'll have to reimplement them yourself. I think [torchcde](https://github.com/patrick-kidger/torchcde) does have an implementation you can use as a starting but IIRC this implementation is mistakenly noncausal in the presence of missing data (represented as NaNs). More broadly if you want a paper reference for backward Hermite cubic splines btw then see [here](https://arxiv.org/abs/2106.11028).

**Comment by Trotew** [comment]
Really appreciate your help.

I think I need to spend some time researching this problem.

**Comment by yangling-bit** [comment]
Hello,
I have the same question and would like to ask if you have any code implementation or could recommend any relevant library functions?
